geoAssembler/widgets/qt_objects.py

```diff
@@ -158,36 +158,45 @@ def _apply(self):
         self.filename_set_signal.emit()
         self.close()
 
-class QLogger(logging.Handler):
-    """Logger object connected python logging."""
 
-    def __init__(self, main_widget):
-        """Create a Dialog that displays the log connected to logging.
+class LogCapturer(logging.Handler, QtCore.QObject):
+    """Emit Qt signal for Python logging, and store recent messages.
+    """
+    new_msg = QtCore.Signal(str)
 
-        Parameters:
-            main_widget : Parent creating this dialog
-        """
-        super().__init__()
-        self.win = QtGui.QDialog(main_widget)
-        layout = QtWidgets.QGridLayout()
-
-        self.widget = QtGui.QPlainTextEdit()
-        self.widget.setReadOnly(True)
-        layout.addWidget(self.widget, 0, 0, 10, 10)
-        ok_btn = create_button('Ok', 'ok')
-        ok_btn.clicked.connect(lambda: self.win.close())
-        layout.addWidget(ok_btn, 11, 0, 1, 1)
-        self.win.setLayout(layout)
-
-    def show(self):
-        """Show the log window."""
-        self.win.show()
+    def __init__(self, parent, level=logging.NOTSET):
+        QtCore.QObject.__init__(self, parent)
+        logging.Handler.__init__(self, level)
+        self.msg_buffer = deque(maxlen=500)
 
     def emit(self, record):
-        """Overload emit signal to write into text widget."""
         msg = self.format(record)
-        self.widget.appendPlainText(msg)
+        self.msg_buffer.append(msg)
+        self.new_msg.emit(msg)
+
+
+class LogDialog(QtGui.QDialog):
+    """A Dialog that displays the log connected to logging.
+
+    Parameters:
+        main_widget : Parent creating this dialog
+    """
+    def __init__(self, main_window):
+        super().__init__(parent=main_window)
+        self.setAttribute(QtCore.Qt.WA_DeleteOnClose)
+        self.log_capturer = main_window.log_capturer
+
+        self.text_area = QtGui.QPlainTextEdit(self)
+        self.text_area.setReadOnly(True)
+        self.ok_btn = create_button('Ok', 'ok')
+        self.ok_btn.clicked.connect(self.close)
+
+        layout = QtWidgets.QGridLayout(self)
+        layout.addWidget(self.text_area, 0, 0, 10, 10)
+        layout.addWidget(self.ok_btn, 11, 0, 1, 1)
+        self.setLayout(layout)
+
+        for msg in self.log_capturer.msg_buffer:
+            self.text_area.appendPlainText(msg)
 
-    def write(self, m):
-        """Overload write and do nothing."""
-        pass
+        self.log_capturer.new_msg.connect(self.text_area.appendPlainText)
```

Comment:
I guess this is for the logging window to display new messages while it is still open? At least on my desktop environment this has no effect because the main window is deactivated until the logger window is closed. In other words you can't trigger any new log messages while this window is open.

Reply:
That is the idea. I noticed the dialog is modal (i.e. blocks the main window) at the moment, but maybe at some point it won't be, or there might be long running operations which you could start and then open the logging dialog to check their progress, so I thought it was worth connecting it.